src/python/CRABClient/Commands/setdatasetstatus.py

```diff
@@ -0,0 +1,118 @@
+# pylint: disable=consider-using-f-string, unspecified-encoding
+"""
+allow users to (in)validate their own DBS USER datasets
+"""
+
+import sys
+import json
+
+from CRABClient.Commands.SubCommand import SubCommand
+from CRABClient.ClientExceptions import MissingOptionException, ConfigurationException, CommandFailedException
+from CRABClient.ClientUtilities import colors
+from CRABClient.RestInterfaces import getDbsREST
+
+if sys.version_info >= (3, 0):
+    from urllib.parse import urlencode  # pylint: disable=E0611
+if sys.version_info < (3, 0):
+    from urllib import urlencode
+
+
+class setdatasetstatus(SubCommand):
+    """
+    Set status of a USER dataet in phys03,
+    optionally invalidates/revalidates all files in it
```

Comment:
I have a question: does this `optionally` means that it is possible to have valid files in an invalid dataset?
I mean, i have a valid dataset with valid files. When happens when i call `setdatasetstatus` without the `--recursive` argument? only the dataset is invalidated, or the dataset and all the files in it?

Reply:
dataset status and file status are independent concepts iin DBS.
Files only have a boolean flag: is_file_valid=0/1
In practice this is used to mark files lost or otherwise unreadable.

Dataset status is a flag which can have many values and it is used by P&R also to track status, e.g. PRODUCTION means the dataset is sitll being populated. Other important values are:
DEPRECATED (e.g. there is newer version, but it could still be used e.g. to compare)
INVALID (contents are known to be no good for physics, bugs in the code, bad conditions.. but again there can be reasons to look at data)
DELETED (the name says it all)
In CRAB we always told users to only use VALID/INVALID for datasets, to avoid creating confusion. In the end if you flag your USER dataset as INVALID, nobody should try to read w/o checking with you.

So in a way is_file_valid is only meaningful for non-DELETED datasets, and its function is to allow data processors to skip missing files. We never delete records from DBS and we never delete the info that a file was part of the dataset at some time, simply record that it became unreadable (and should it happend that it gets rescued, we can still turn it back into valid).

In theory that's all and there is no need to invalidate all files in a dataset.
In practive, few people care to know the full story, and it is not written anywhere, and P&R may change the meaning of the dataset status as they please, so it makes some sense that if you delete your USER dataset from disk, you also want to mark all files with  `is_file_valid=0`, just to be on the safe side.

By default the files status is not affected by set dataset status.
And the idea for --recursive is to flatly clean everything by marking as invalid all files in the dataset.
It is not likely that users have children datasets to worry about. You may argue that you want to delete the parent and keep childs around, so after the --recursive invalidation there will be a need for a few validation.

Note that I simply tried to reproduce the semantic of old scripts which were developed (long time ago) for production operators in which case marking children datasets as e.g. INVALID for important. I am not even sure that at that time operators really knew the difference between DELETED and INVALID. As I said nothing was ever written down and the fact that dataset validity is a string, and file validity one bit, makes it difficult to have consistency w/o strong conventions.  Ref: https://github.com/dmwm/DBS/blob/master/Client/utils/DataOpsScripts/DBS3SetDatasetStatus.py
Indeed I failed ! Since the original [would mark](https://github.com/dmwm/DBS/blob/14df8bbe8ee8f874fe423399b18afef911fe78c7/Client/utils/DataOpsScripts/DBS3SetDatasetStatus.py#L59) all files in a dataset as invalid if dataset type is no VALID. IMO that's bad. And I hope operators have not used this exact script, in the end (indeed only CRAB users complained when it stopped working).

That said, since --recursive is not implemented yet, we can do what we want. I am personally in favor of doing w/o it. Indeed the above scenario is a good part of why I decided to postpone it !

We can  clarify in the instructions that it is up to users to mark all files as well (pretty easy anyhow) and look for childrens (we can also put instructions e.g. [1]). In the end this `--recursive` in the original was confusing
and clearly ad hoc for some specific use case. 

Bottom lines, users should do:
1. I lost a file on disk: crab setfilestuatus  --file ... --status INVALID (once for every file until dbs2go bug is fixed)
2. I want to make sure nobody tries to read this (e.g. gone from disk):
   * crab setdataset --dataset ... --status INVALID (or DELETED)
   * AND TO BE FULLY SURE (since you can ask CRAB e.g. to read a dataset in any status):
   * crab setfilestatus --dataset ... --status INVALID



[1]
```
 belforte@lxplus805/TC3> curl --cert $X509_USER_PROXY  --key $X509_USER_PROXY -H "Accept: application/json" "https://cmsweb.cern.ch:8443/dbs/prod/global/DBSReader/datasetchildren?dataset=/ZeroBias5/Run2023F-v1/RAW"
[
{"child_dataset":"/ZeroBias5/Run2023F-PromptReco-v1/DQMIO","child_dataset_id":14767495,"dataset":"/ZeroBias5/Run2023F-v1/RAW"}
,{"child_dataset":"/ZeroBias5/Run2023F-LogErrorMonitor-PromptReco-v1/USER","child_dataset_id":14767497,"dataset":"/ZeroBias5/Run2023F-v1/RAW"}
,{"child_dataset":"/ZeroBias5/Run2023F-LogError-PromptReco-v1/RAW-RECO","child_dataset_id":14767500,"dataset":"/ZeroBias5/Run2023F-v1/RAW"}
,{"child_dataset":"/ZeroBias5/Run2023F-HcalCalIsolatedBunchSelector-PromptReco-v1/ALCARECO","child_dataset_id":14767504,"dataset":"/ZeroBias5/Run2023F-v1/RAW"}
,{"child_dataset":"/ZeroBias5/Run2023F-SiStripCalMinBias-PromptReco-v1/ALCARECO","child_dataset_id":14767508,"dataset":"/ZeroBias5/Run2023F-v1/RAW"}
,{"child_dataset":"/ZeroBias5/Run2023F-SiStripCalZeroBias-PromptReco-v1/ALCARECO","child_dataset_id":14767509,"dataset":"/ZeroBias5/Run2023F-v1/RAW"}
,{"child_dataset":"/ZeroBias5/Run2023F-TkAlMinBias-PromptReco-v1/ALCARECO","child_dataset_id":14767519,"dataset":"/ZeroBias5/Run2023F-v1/RAW"}
,{"child_dataset":"/ZeroBias5/Run2023F-PromptReco-v1/NANOAOD","child_dataset_id":14767525,"dataset":"/ZeroBias5/Run2023F-v1/RAW"}
,{"child_dataset":"/ZeroBias5/Run2023F-PromptReco-v1/MINIAOD","child_dataset_id":14767544,"dataset":"/ZeroBias5/Run2023F-v1/RAW"}
,{"child_dataset":"/ZeroBias5/Run2023F-PromptReco-v1/AOD","child_dataset_id":14767549,"dataset":"/ZeroBias5/Run2023F-v1/RAW"}
]
belforte@lxplus805/TC3> 
```